### ISSUE TITLE
Fasten 'vak prep predict.toml'

### ISSUE THREAD

**Issue by fMizki** [vocalpy/vak#507]
I'm a student in  Prof. Abe's lab in Tohoku University.
Recently I'm working with your great 'Tweetynet' and thus 'vak'.
Though it is not a very big matter, when I run  'vak prep *(predict).toml' in the prediction phase, it takes me a long time as I give so many files. When I gave 50000 files for a prediction, it took me about 13h for 'vak prep'.
Is there any way to shorten this consuming time?

I thought it might help me if I could allocate several CPUs for several processing or use GPU for this calculation, but I'm not skilled enough yet. So I'm grad if you could help me. 

Please ignore it if it bothers you, as I can treat it even if it costs time, anyway.
Thank, you.

**Comment by NickleDave** [comment]
Hi @fMizki !

> I'm a student in Prof. Abe's lab in Tohoku University.
Recently I'm working with your great 'Tweetynet' and thus 'vak'.

Great, glad to hear it and happy to help

> Though it is not a very big matter, when I run 'vak prep *(predict).toml' in the prediction phase, it takes me a long time as I give so many files. When I gave 50000 files for a prediction, it took me about 13h for 'vak prep'.
Is there any way to shorten this consuming time?

Just to make sure I understand:  
the step that takes a long time right now is `prep`, *not* `predict`, right?  
I.e., you run `vak prep` and it generates spectrograms from your (50k!) files?  This is what's taking 13h?  

If so, then the short answer is:  
right now there's (probably) not an easy way to speed up this step.  

The longer answer is:  
`vak` uses the [`dask` library](https://dask.org/) to parallelize parts of the `prep` step, like generating the spectrogram files.  
`dask` *should* be using as many cores as possible on your machine.

If you are on a UNIX system (Mac OS / Linux) you could check whether this is the case by running the program `htop` in another terminal while `vak prep` is running. You should see that all the cores are being used. I'm not sure how to check on Windows but I could try to find out. One option might be to use the Windows Subsystem for Linux (WSL) and run `htop` from a terminal running in WSL. (I can try to explain more what I mean by that if you are on Windows.)

If it's easy for you to switch to a machine with more cores, and you are still able to access the data on that machine to run the `predict` step, then you might find that you get a speed up by simply switching to a more powerful machine.

This is the main function where generating the spectrogram files with `dask` takes place, in case you want to get an idea of what's going on:  
https://github.com/vocalpy/vak/blob/main/src/vak/io/spect.py

Specifically, this line is probably the main step that takes up most of the time during that 13 hours, I'd guess:  
https://github.com/vocalpy/vak/blob/10a6ea23715339e17d9b4d9a944c708fcfa18167/src/vak/io/spect.py#L267

Generating spectrograms on the GPU is definitely something we have thought about but there's no easy way to do that right now.  

One other question:  
I guess you are running a behavioral experiment, and that's how you ended up with 50k files?  
A quick-and-dirty fix might be to set up `vak prep` to run automatically somehow, e.g. once every day on any new folders you add to some "sorted song files" directory.
For example you could set up a [cron job script](https://www.hostinger.com/tutorials/cron-job).  
This doesn't make `vak` any faster but at least you can stay up-to-date instead of having to run on all the files at one time.  

Please let me know if that helps. Happy to talk more.

**Comment by fMizki** [comment]
Thank you for your reply.

> Just to make sure I understand:
the step that takes a long time right now is prep, not predict, right?
I.e., you run vak prep and it generates spectrograms from your (50k!) files? This is what's taking 13h?

Yes, that's what I meant.

So, do you think the total CPU usage reaches nearly100% in an ideal condition?
Well, I'm on Windows not on a UNIX system (Mac OS / Linux) because when I tried on WSL, I got `segment fault` error even when I commanded `vak --help` and I couldn't solve this problem. On Windows, I think we can get similar information to `htop` with "Windows Task Manager", but CPU usage was less than 16% in total and less than10% for python(6 cores computer). I wonder it might be partly because of the high disk usage which was about 100%, though I don't understand the consumer of it.
Actually, we are to have very powerful machine, so I'll try on it.

> One other question:
I guess you are running a behavioral experiment, and that's how you ended up with 50k files?

Yes I get about 50k files for one period of my behavioral experiment. 
You are right. I would have better run `vak` more frequently. As I am using different computer for recording and analysis, transferring files with USB HD is what hinders me from doing it. I'm afraid of the loss of data when I access them from different PC. So I might choose to run by my self without scheduling when I move files, but anyway the best way is staying up-to-date.

I'm enthusiastic to learn about computer programing and get more skilled, so I'm really enjoying to talk with you and I appreciate for you to tell me things like `dask` and "cron job script" that I didn't know and I wanted sometimes for my work. They must help me.

**Comment by NickleDave** [comment]
> So, do you think the total CPU usage reaches nearly100% in an ideal condition? 

I'm not sure about 100% but as a quick check in the past I have run `htop` just to make sure more than one core is being used--that's all I mean.

> Well, I'm on Windows not on a UNIX system (Mac OS / Linux) because when I tried on WSL, I got `segment fault` error even when I commanded `vak --help` and I couldn't solve this problem. 

Hmm, not sure what's going on with the segfault.  
WSL can be tricky.  

> On Windows, I think we can get similar information to `htop` with "Windows Task Manager", but CPU usage was less than 16% in total and less than10% for python(6 cores computer).

You did see that more than one core was being used then?  
If so then we at least know `dask` is doing what we expect it to

> I wonder it might be partly because of the high disk usage which was about 100%, though I don't understand the consumer of it. Actually, we are to have very powerful machine, so I'll try on it.

That might be your best bet for now, unfortunately.

> > One other question:
> > I guess you are running a behavioral experiment, and that's how you ended up with 50k files?
> 
> Yes I get about 50k files for one period of my behavioral experiment. You are right. I would have better run `vak` more frequently. As I am using different computer for recording and analysis, transferring files with USB HD is what hinders me from doing it. I'm afraid of the loss of data when I access them from different PC. So I might choose to run by my self without scheduling when I move files, but anyway the best way is staying up-to-date.

Probably.  
Are those 50k files filtered in any way?  
E.g., if you have a recorder that is triggered any time that amplitude of a mic goes above some threshold, you might end up with a lot of false positives, e.g. because of noises like perch movements. Do you have a clean-up step that tries to remove those false positives?

**Comment by NickleDave** [comment]
> I'm enthusiastic to learn about computer programing and get more skilled, so I'm really enjoying to talk with you and I appreciate for you to tell me things like `dask` and "cron job script" that I didn't know and I wanted sometimes for my work. They must help me.

Happy to help.

@fMizki can you share an email with me?  
We're setting up a forum for people to ask questions like yours here, I can send you an invite.
If you don't feel comfortable posting it here you can send it to my email: nicholdav at gmail dot com

**Comment by fMizki** [comment]
>  I'm not sure about 100% but as a quick check in the past I have run htop just to make sure more than one core is being used--that's all I mean.

I see. Then I'm not sure I can check how many cores are allocated to each task or which core is allocated with  "Windows Task Manager", so I'll find other ways.

>>  Well, I'm on Windows not on a UNIX system (Mac OS / Linux) because when I tried on WSL, I got segment fault error even when I commanded vak --help and I couldn't solve this problem.

> Hmm, not sure what's going on with the segfault.
WSL can be tricky.

Just for information, actually, another member of us also encountered `segment fault(core dumped)` on VBox after `simple-seq` introduction, whose environment had been working before it, though I can't say our causes of errors are same.

> Probably.
Are those 50k files filtered in any way?
E.g., if you have a recorder that is triggered any time that amplitude of a mic goes above some threshold, you might end up with a lot of false positives, e.g. because of noises like perch movements. Do you have a clean-up step that tries to remove those false positives?

I don't filter files now, as I don't come up with a good way.  I discard unneeded files after interpretation based on its result, e.g. the variety of morphemes( or morpheme like noises) and density of them, and use the rest for analysis. 
I'm sure about half of my 50k files are noises or calls. Discarding based on their file size might be one way, but I'm not sure how accurately it remove only noise and call files and commonly it is accepted. Do you have some ideas for this clean-up step?

> @fMizki can you share an email with me?
We're setting up a forum for people to ask questions like yours here, I can send you an invite.
If you don't feel comfortable posting it here you can send it to my email: nicholdav at gmail dot com

Thank you! I'll send you an email.

**Comment by NickleDave** [comment]
> Just for information, actually, another member of us also encountered segment fault(core dumped) on VBox after simple-seq introduction, whose environment had been working before it, though I can't say our causes of errors are same.

This is helpful to know, thank you.

Segfaults are definitely not easy to troubleshoot.

Does it happen in a totally new environment, or is it after upgrading some packages (e.g., a new version of `vak` with the `'simple-seq'` format added)? If it happens after upgrading, does it fix things to totally create a new environment from scratch? You might try keeping the old environment around and running in both a couple of times to see if you only get the segfault in the old one but not the new one. I don't actually know how we would fix things if this is a recurrent issue but starting over with a new environment "from scratch" is the first thing I would do if I were trying to fix it.

**Comment by NickleDave** [comment]
> I don't filter files now, as I don't come up with a good way. I discard unneeded files after interpretation based on its result, e.g. the variety of morphemes( or morpheme like noises) and density of them, and use the rest for analysis.
I'm sure about half of my 50k files are noises or calls. Discarding based on their file size might be one way, but I'm not sure how accurately it remove only noise and call files and commonly it is accepted. Do you have some ideas for this clean-up step?

We have some scripts we used in Sober lab.  
I looked at them but they rely so much on other code that I'm not sure it would help you to share them.  
The basic idea is:  
- first check the ratio of power spectral density above and below 10k
  - for an isolated bird in a sound box, it seems to be the case that when the ratio mean(PSD( > 10k)) / mean(PSD( < 10k)) > n, where n is some parameter you tune (we use 0.1), then this is mostly hopping around noise.
- if that first criteria is met ( ratio < n), then count the number of threshold crossings of amplitude, i.e. segment the audio -- if the number of segments is above some parameter you tune, then keep it. I.e., just two segments is probably a call, but 100 segments is likely to be a song.

I'm sorry, I know that's not super helpful.  I just don't want to dump a bunch of code on you that might be even less helpful.

If it were me, now, I would train a little convolutional neural network to classify "song" and "not song" spectrograms, and sort them that way. But you would definitely want to make sure it works with a test dataset (maybe you already have one from what you've sorted by hand?). I know that might not be super helpful either

I tried to dig around for some other songbird labs' code but didn't find anything there
https://github.com/soberlab
https://github.com/gardner-lab
https://github.com/melizalab
https://github.com/melizalab
(don't have too much fun distracting yourself :smirk: )

**Comment by NickleDave** [comment]
> We got a new powerful computer with RTX3090, which is supported with "cuda" of which version is more than 11.0 and thus vak's "pytorch" version seems unmatched, so I get the warning;

No worries. I'm happy to help and I think I know the fix to this. Can you please do the following:

1. close this issue (press "close" button below), unless you feel like there's something we haven't addressed here about speeding up `vak`
  - I'm happy to discuss more on the forum, but we're trying to move general help questions there, and use GitHub mainly for bug reports and feature requests
2. open up a **new issue** that's a bug report (big green button that says "bug report") where you write it exactly as you did in your comment above
  - it will help us if you make a new issue, since that will make it easier for people to find when they search. Please include specifics in the title, something like "CUDA capability sm_86 error with pytorch"
  - I will reply there with the fix I just gave someone else, and that will help us track if what I told them is a general way to resolve the problem for everyone
  - if so we might need to add it to the docs, e.g. on the installation page

Thank you! 🙏

**Comment by fMizki** [comment]
I understand.
Thank you for your help!

**Comment by fMizki** [comment]
Problems might be solved.

Working on network drive makes it very slow. Details are on https://github.com/vocalpy/vak/issues/509.
